5. PANDAS BÁSICO

<span style="font-size:14px;">Pandas é uma ferramenta de manipulação de dados de alto nível, construída com base no Numpy. O pacote Pandas possui estruturas de adados bastante interessantes para manipulação de dados e por isso é muito utilizado por cientista de dados. Pandas é uma biblioteca para Python. Ele possui o objetivo principal de criar séries, dataframes e array.
  
<span style="font-size:14px;">No Pandas, objects = string 

In [75]:
import pandas as pd


In [76]:
# Criando uma Séries a partir de uma lista
# Lista:
carros = ['Jetta Variant', 'Passat','Crossfox']
carros


['Jetta Variant', 'Passat', 'Crossfox']

In [77]:
# Séries / traz um índice automático de 0 a n
pd.Series(carros)


0    Jetta Variant
1           Passat
2         Crossfox
dtype: object

Criando um DataFrame a partir de uma LISTA de dicionários

In [78]:
dados = [
    {'Nome': 'Jetta Variant', 'Motor': 'Motor 4.0 Turbo', 'Ano': 2003, 'Quilometragem': 44410.0, 'Zero_km': False, 'Valor': 88078.64},
    {'Nome': 'Passat', 'Motor': 'Motor Diesel', 'Ano': 1991, 'Quilometragem': 5712.0, 'Zero_km': False, 'Valor': 106161.94},
    {'Nome': 'Crossfox', 'Motor': 'Motor Diesel V8', 'Ano': 1990, 'Quilometragem': 37123.0, 'Zero_km': False, 'Valor': 72832.16}
]


In [79]:
# Contruindo o DataFrame

dataset = pd.DataFrame(dados)
dataset


Nome            Motor   Ano  Quilometragem  Zero_km      Valor
0  Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   88078.64
1         Passat     Motor Diesel  1991         5712.0    False  106161.94
2       Crossfox  Motor Diesel V8  1990        37123.0    False   72832.16

In [80]:
# Alterando as coluns de lugar
dataset[['Ano','Nome','Motor','Quilometragem','Valor','Zero_km']]


Ano           Nome            Motor  Quilometragem      Valor  Zero_km
0  2003  Jetta Variant  Motor 4.0 Turbo        44410.0   88078.64    False
1  1991         Passat     Motor Diesel         5712.0  106161.94    False
2  1990       Crossfox  Motor Diesel V8        37123.0   72832.16    False

Criando um DataFrame a partir de um DICIONARIO

In [81]:
# Aqui a chave é o primeiro item da linha e o valor é tudo que está dentro do colchete
dados = {
    'Nome': ['Jetta Variant', 'Passat', 'Crossfox'], 
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8'],
    'Ano': [2003, 1991, 1990],
    'Quilometragem': [44410.0, 5712.0, 37123.0],
    'Zero_km': [False, False, False],
    'Valor': [88078.64, 106161.94, 72832.16]
}


In [82]:
dataset = pd.DataFrame(dados)
dataset


Nome            Motor   Ano  Quilometragem  Zero_km      Valor
0  Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   88078.64
1         Passat     Motor Diesel  1991         5712.0    False  106161.94
2       Crossfox  Motor Diesel V8  1990        37123.0    False   72832.16

Criando um DataFrame a partir de um arquivo externo

In [83]:
dataset = pd.read_csv(r"C:\Users\fabia\OneDrive\FIAP_INICIO\WarnUp_materiais\Python_Data_Science\Pandas\data\db.csv", sep = ';', index_col= 0 )
# estou declarando que quero que a minha primeira coluna seja o próprio indice / para não vir a coluna de índice automático


In [84]:
dataset


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor  
Nome                              
Jetta Variant           88078.64  
Passat                 106161.94  
Crossfox                72832.16  
DS5                    124549.07  
Aston Martin DB4        92612.10  
...                          ...  
Phantom 2013            51759.58  
Cadillac Ciel concept   51667.06  
Classe GLK              68934.03  
Aston Martin DB5       122110.90  
Macan                   90381.47  

[258 rows x 6 columns]

EXERCICIOS:

<span style="font-size:14px;">Vimos no último vídeo que é possível criar DataFrames e Series a partir de várias fontes (arquivos externos, listas, dicionários, etc).  
  
<span style="font-size:14px;">Na aula que falamos sobre funções, desenvolvemos em nossos exercícios uma função que recebia um dicionário com um conjunto de informações sobre veículos, e calculava a quilometragem média anual de cada veículo. Esta função retornava o conteúdo do dicionário de input da função, incluindo as informações sobre a quilometragem média.  
  
<span style="font-size:14px;">Dica: Para resolver esta questão, será necessário relembrar de um recurso que aprendemos no curso anterior, quando falamos de arrays Numpy. Este recurso também pode ser aplicado a DataFrames do pandas:  
  
<span style="font-size:14px;">ndarray.T: Retorna o array transposto, isto é, converte linhas em colunas e vice versa.

In [85]:
dados = {
    'Crossfox': {'km': 35000, 'ano': 2005}, 
    'DS5': {'km': 17000, 'ano': 2015}, 
    'Fusca': {'km': 130000, 'ano': 1979}, 
    'Jetta': {'km': 56000, 'ano': 2011}, 
    'Passat': {'km': 62000, 'ano': 1999}
}


In [86]:
def km_media(dataset, ano_atual):
    result={} # linha 3 a 6 é tudo o que eu devo ter dentro do result / resultado desta função
    for item in dataset.items(): # para cada item/linha do dataset terei:
        media = item[1]['km'] / (ano_atual - item[1]['ano']) #media de quilomentragem,
        item[1].update({'km_media': media}) #estou acrescentando esta informação no dicionário / , a informação (nova) 'km_media' e o valor da quilometragem da média (linha4)
        result.update({item[0]:item[1]}) # a informação acima será alocada no item [0] = chave e no item [1] = valores correspondentes as chaves
    return result


In [87]:
km_media(dados, 2019)


{'Crossfox': {'km': 35000, 'ano': 2005, 'km_media': 2500.0},
 'DS5': {'km': 17000, 'ano': 2015, 'km_media': 4250.0},
 'Fusca': {'km': 130000, 'ano': 1979, 'km_media': 3250.0},
 'Jetta': {'km': 56000, 'ano': 2011, 'km_media': 7000.0},
 'Passat': {'km': 62000, 'ano': 1999, 'km_media': 3100.0}}

In [88]:
# Construindo um DataFrame, a partir do dicionario acima, onde a coluna chave seja os nomes dos carros e values/atributos sejam o que esta 
#dentro do dicionário (km, ano e km_media)
carros = pd.DataFrame(km_media(dados, 2019))


In [89]:
carros


Crossfox      DS5     Fusca    Jetta   Passat
km         35000.0  17000.0  130000.0  56000.0  62000.0
ano         2005.0   2015.0    1979.0   2011.0   1999.0
km_media    2500.0   4250.0    3250.0   7000.0   3100.0

In [90]:
# Transpondo o DataFrame, basta colocar o T no final do código
carros = pd.DataFrame(km_media(dados, 2019)).T


In [91]:
carros


km     ano  km_media
Crossfox   35000.0  2005.0    2500.0
DS5        17000.0  2015.0    4250.0
Fusca     130000.0  1979.0    3250.0
Jetta      56000.0  2011.0    7000.0
Passat     62000.0  1999.0    3100.0

5.2 Seleções com DataFrames

In [92]:
dataset.head() # traz as 5 primeiras linhas do DataFrame


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

In [93]:
# Selecionando Colunas --> Valor / Traz uma Series do Pandas

dataset['Valor']


Nome
Jetta Variant             88078.64
Passat                   106161.94
Crossfox                  72832.16
DS5                      124549.07
Aston Martin DB4          92612.10
                           ...    
Phantom 2013              51759.58
Cadillac Ciel concept     51667.06
Classe GLK                68934.03
Aston Martin DB5         122110.90
Macan                     90381.47
Name: Valor, Length: 258, dtype: float64

In [94]:
# para saber o tipo do Dataset
type(dataset['Valor']) 


pandas.core.series.Series

In [95]:
# Se eu quiser que o resultado me retorne um DataFrame: colocar 2 colchetes

dataset[['Valor']]



Valor
Nome                            
Jetta Variant           88078.64
Passat                 106161.94
Crossfox                72832.16
DS5                    124549.07
Aston Martin DB4        92612.10
...                          ...
Phantom 2013            51759.58
Cadillac Ciel concept   51667.06
Classe GLK              68934.03
Aston Martin DB5       122110.90
Macan                   90381.47

[258 rows x 1 columns]

In [96]:
type(dataset[['Valor']])


pandas.core.frame.DataFrame

Selecionando Linhas - [i:j]  
  
Obeservação: A indexação tem origem no zero e nos fatiamentos (slices) a linha com índice i é incluída e a linha com índice j não é incluída no resultado

In [97]:
# Selecionando linhas / não conta cabeçalho / e a linha 3 do dataframe não vem no resultado
dataset[0:3]


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                           
Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   
Passat            Motor Diesel  1991         5712.0    False   
Crossfox       Motor Diesel V8  1990        37123.0    False   

                                                      Acessórios      Valor  
Nome                                                                         
Jetta Variant  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat         ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox       ['Piloto automático', 'Controle de estabilidad...   72832.16

LOC: Utilizando .loc para seleções

In [98]:
dataset.loc['Passat'] # Retorna uma lista referente a linha do Passat


Motor                                                 Motor Diesel
Ano                                                           1991
Quilometragem                                               5712.0
Zero_km                                                      False
Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
Valor                                                    106161.94
Name: Passat, dtype: object

In [99]:
# Se quisermos mais de uma informação --> usar 2 colchetes / retorna um DataFrame

dataset.loc[['Passat', 'DS5']]


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                    
Passat     Motor Diesel  1991         5712.0    False   
DS5     Motor 2.4 Turbo  2019            NaN     True   

                                               Acessórios      Valor  
Nome                                                                  
Passat  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
DS5     ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

O .loc também nos permite selecionar quais colunas desejamos exibir. Para isso, depois das informações das linhas, adicionaremos uma vírgula e um novo par de colchetes, dessa vez com as informações "Motor" e "Valor" representando as colunas a serem exibidas.

In [100]:
dataset.loc[['Passat','DS5'],['Motor','Valor']]


Motor      Valor
Nome                              
Passat     Motor Diesel  106161.94
DS5     Motor 2.4 Turbo  124549.07

Já se passarmos apenas : como primeiro parâmetro, receberemos todas as linhas do conjunto e somente as colunas selecionadas.

In [101]:
dataset.loc[:,['Motor','Valor']]


Motor      Valor
Nome                                                   
Jetta Variant                Motor 4.0 Turbo   88078.64
Passat                          Motor Diesel  106161.94
Crossfox                     Motor Diesel V8   72832.16
DS5                          Motor 2.4 Turbo  124549.07
Aston Martin DB4             Motor 2.4 Turbo   92612.10
...                                      ...        ...
Phantom 2013                        Motor V8   51759.58
Cadillac Ciel concept               Motor V8   51667.06
Classe GLK             Motor 5.0 V8 Bi-Turbo   68934.03
Aston Martin DB5                Motor Diesel  122110.90
Macan                        Motor Diesel V6   90381.47

[258 rows x 2 columns]

ILOC: O iloc também nos permite fazer seleções, mas se utiliza dos índices numéricos - ou seja, na posição das informações. No primeiro exemplo, passaremos como parâmetro apenas 1.

In [102]:
dataset.iloc[1]


Motor                                                 Motor Diesel
Ano                                                           1991
Quilometragem                                               5712.0
Zero_km                                                      False
Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
Valor                                                    106161.94
Name: Passat, dtype: object

In [103]:
dataset.iloc[['1']]


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                 
Passat  Motor Diesel  1991         5712.0    False   

                                               Acessórios      Valor  
Nome                                                                  
Passat  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94

Quando queremos informações de múltiplas linhas em um dataset, utilizaremos somente um par de colchetes, dentro dos quais passaremos o índice inicial e o final (lembrando que o inicial é inclusivo e o final não).

In [104]:
dataset.iloc[1:4]


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                      
Passat       Motor Diesel  1991         5712.0    False   
Crossfox  Motor Diesel V8  1990        37123.0    False   
DS5       Motor 2.4 Turbo  2019            NaN     True   

                                                 Acessórios      Valor  
Nome                                                                    
Passat    ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox  ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5       ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

In [105]:
type(dataset.iloc[1:4])


pandas.core.frame.DataFrame

Para obtermos colunas específicas, passaremos, após uma vírgula, um novo par de colchetes contendo os índices das colunas que desejamos acessar. Não é necessário passá-los na mesma ordem do dataset original, algo que exemplificaremos na seleção abaixo, feita com as colunas 0, 5 e 2 ("Motor", "Valor" e "Quilometragem", respectivamente).

In [106]:
dataset.iloc[1:4,[0,5,2]]


Motor      Valor  Quilometragem
Nome                                               
Passat       Motor Diesel  106161.94         5712.0
Crossfox  Motor Diesel V8   72832.16        37123.0
DS5       Motor 2.4 Turbo  124549.07            NaN

Da mesma forma, é possível selecionar linhas específicas que não estejam em sequência, bastando passarmos seus índices dentro de um par de colchetes.

In [107]:
dataset.iloc[[1,42,22],[0,5,2]]


Motor      Valor  Quilometragem
Nome                                                          
Passat                  Motor Diesel  106161.94         5712.0
Optima                 Motor 1.8 16v   86641.34            NaN
Lamborghini Obvious  Motor Diesel V6  133529.84        98079.0

Se quisermos exibir todas as linhas, usaremos os dois pontos (:).

In [108]:
dataset.iloc[:,[0,5,2]]


Motor      Valor  Quilometragem
Nome                                                                  
Jetta Variant                Motor 4.0 Turbo   88078.64        44410.0
Passat                          Motor Diesel  106161.94         5712.0
Crossfox                     Motor Diesel V8   72832.16        37123.0
DS5                          Motor 2.4 Turbo  124549.07            NaN
Aston Martin DB4             Motor 2.4 Turbo   92612.10        25757.0
...                                      ...        ...            ...
Phantom 2013                        Motor V8   51759.58        27505.0
Cadillac Ciel concept               Motor V8   51667.06        29981.0
Classe GLK             Motor 5.0 V8 Bi-Turbo   68934.03        52637.0
Aston Martin DB5                Motor Diesel  122110.90         7685.0
Macan                        Motor Diesel V6   90381.47        50188.0

[258 rows x 3 columns]

6. Fatiamento com DataFrames

In [109]:
# Considere o seguinte código:
import pandas as pd

dados = {
    'Nome': ['Jetta', 'Passat', 'Crossfox', 'DS5', 'Fusca'], 
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8', 'Motor 2.0', 'Motor 1.6'],
    'Ano': [2019, 2003, 1991, 2019, 1990],
    'Quilometragem': [0.0, 5712.0, 37123.0, 0.0, 120000.0],
    'Zero_km': [True, False, False, True, False],
    'Valor': [88000.0, 106000.0, 72000.0, 89000.0, 32000.0]
}

datasetex = pd.DataFrame(dados)


Assinale a alternativa que contenha o código que seleciona somente as informações de Nome, Ano, Quilometragem e Valor dos carros Passat e Crossfox.

In [110]:
datasetex[['Nome','Ano','Quilometragem','Valor']][1:3]


Nome   Ano  Quilometragem     Valor
1    Passat  2003         5712.0  106000.0
2  Crossfox  1991        37123.0   72000.0

Utilize o código abaixo para resolver a questão:  
  
Note que utilizamos os nomes dos veículos como índice do DataFrame. Selecione as alternativas corretas que têm como resultado o um DataFrame:

In [111]:
import pandas as pd

dados = {
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8', 'Motor 2.0', 'Motor 1.6'],
    'Ano': [2019, 2003, 1991, 2019, 1990],
    'Quilometragem': [0.0, 5712.0, 37123.0, 0.0, 120000.0],
    'Zero_km': [True, False, False, True, False],
    'Valor': [88000.0, 106000.0, 72000.0, 89000.0, 32000.0]
}

datasetex = pd.DataFrame(dados, index = ['Jetta', 'Passat', 'Crossfox', 'DS5', 'Fusca'])


In [112]:
datasetex.loc[['Passat','DS5']]


Motor   Ano  Quilometragem  Zero_km     Valor
Passat  Motor Diesel  2003         5712.0    False  106000.0
DS5        Motor 2.0  2019            0.0     True   89000.0

In [113]:
datasetex.loc[['Passat','DS5'],['Motor','Valor']]


Motor     Valor
Passat  Motor Diesel  106000.0
DS5        Motor 2.0   89000.0

In [114]:
# ou poderia também ser utilizado a função iloc

datasetex.head()


Motor   Ano  Quilometragem  Zero_km     Valor
Jetta     Motor 4.0 Turbo  2019            0.0     True   88000.0
Passat       Motor Diesel  2003         5712.0    False  106000.0
Crossfox  Motor Diesel V8  1991        37123.0    False   72000.0
DS5             Motor 2.0  2019            0.0     True   89000.0
Fusca           Motor 1.6  1990       120000.0    False   32000.0

In [115]:
datasetex.iloc[[1,3],[0,4]]


Motor     Valor
Passat  Motor Diesel  106000.0
DS5        Motor 2.0   89000.0

Queries com DataFrames  
  
... que são consultas que podem ser feitas no DataFrame  
  
Queremos realizar uma query que nos traga todos os registros para determinado tipo de motor, por exemplo o "Motor Diesel". Antes de fazermos isso, é interessante comentar que existe outra forma de obtermos uma series simples a partir do rótulo de um dataframe, que é a sintaxe dataframe.Rotulo, como no exemplo abaixo:

In [116]:
dataset.head()


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

In [117]:
# Selecionando uma coluna
dataset.Motor


Nome
Jetta Variant                  Motor 4.0 Turbo
Passat                            Motor Diesel
Crossfox                       Motor Diesel V8
DS5                            Motor 2.4 Turbo
Aston Martin DB4               Motor 2.4 Turbo
                                 ...          
Phantom 2013                          Motor V8
Cadillac Ciel concept                 Motor V8
Classe GLK               Motor 5.0 V8 Bi-Turbo
Aston Martin DB5                  Motor Diesel
Macan                          Motor Diesel V6
Name: Motor, Length: 258, dtype: object

Observação: Em Python, == é um operador de comparação usado para verificar se dois valores são iguais. Ele retorna True se os valores comparados forem iguais e False caso contrário.

In [118]:
# Quero que me retorne todos os carros que são 'Motor Diesel' / Vai me retornar resultado boleano (True or False)
dataset.Motor == 'Motor Diesel'


Nome
Jetta Variant            False
Passat                    True
Crossfox                 False
DS5                      False
Aston Martin DB4         False
                         ...  
Phantom 2013             False
Cadillac Ciel concept    False
Classe GLK               False
Aston Martin DB5          True
Macan                    False
Name: Motor, Length: 258, dtype: bool

In [119]:
# Essa series pode ser passada como parâmetro em uma seleção de dataframe para obtermos somente os registros aos quais o valor "True" foi atribuído. 
# Para facilitarmos esse tipo de consulta, armazenaremos nossa series em uma variável select.
select = dataset.Motor == 'Motor Diesel'


In [120]:
type(select)


pandas.core.series.Series

In [121]:
dataset[select]


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                
Passat                 Motor Diesel  1991         5712.0    False   
Effa Hafei Picape Baú  Motor Diesel  1991       102959.0    False   
Sorento                Motor Diesel  2019            NaN     True   
New Fiesta Hatch       Motor Diesel  2017       118895.0    False   
Kangoo Express         Motor Diesel  2007        29132.0    False   
Fit                    Motor Diesel  2013        44329.0    False   
Cielo Hatch            Motor Diesel  2019            NaN     True   
Symbol                 Motor Diesel  2016       117714.0    False   
A4 Sedan               Motor Diesel  2002        30511.0    False   
A4 Avant               Motor Diesel  2014        17357.0    False   
Silver Shadow          Motor Diesel  2015        99052.0    False   
Camry                  Motor Diesel  2019            NaN     True   
Clio                   Motor Diesel  1990        66437.0    False   
Série 3 M3 Coupé       Motor Diesel  2013        19896.0    False   
Gallardo LP 560 – 4    Motor Diesel  2000       113045.0    False   
Chana Utility          Motor Diesel  2009        52034.0    False   
TT Roadster            Motor Diesel  1998       107392.0    False   
Aston Martin Virage    Motor Diesel  2019            NaN     True   
Sandero                Motor Diesel  2015        34783.0    False   
XKR                    Motor Diesel  1996        85127.0    False   
X5                     Motor Diesel  2002        13606.0    False   
Dodge Dakota           Motor Diesel  1993        71544.0    False   
i30 CW                 Motor Diesel  2015         8497.0    False   
Série 7 Sedã           Motor Diesel  2019            NaN     True   
V60                    Motor Diesel  2004        91840.0    False   
Aston Martin DB5       Motor Diesel  1996         7685.0    False   

                                                              Acessórios  \
Nome                                                                       
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Effa Hafei Picape Baú  ['Controle de estabilidade', 'Painel digital',...   
Sorento                ['Sensor de chuva', 'Câmera de estacionamento'...   
New Fiesta Hatch       ['Sensor de estacionamento', 'Travas elétricas...   
Kangoo Express         ['Bancos de couro', 'Câmbio automático', 'Pilo...   
Fit                    ['Freios ABS', 'Câmera de estacionamento', 'Câ...   
Cielo Hatch            ['Painel digital', 'Central multimídia', 'Câme...   
Symbol                 ['4 X 4', 'Piloto automático', 'Sensor crepusc...   
A4 Sedan               ['Câmera de estacionamento', '4 X 4', 'Travas ...   
A4 Avant               ['Teto panorâmico', '4 X 4', 'Bancos de couro'...   
Silver Shadow          ['4 X 4', 'Central multimídia', 'Ar condiciona...   
Camry                  ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Clio                   ['4 X 4', 'Câmbio automático', 'Freios ABS', '...   
Série 3 M3 Coupé       ['Bancos de couro', 'Sensor de chuva', 'Ar con...   
Gallardo LP 560 – 4    ['Freios ABS', 'Teto panorâmico', 'Central mul...   
Chana Utility          ['Ar condicionado', 'Sensor de estacionamento'...   
TT Roadster            ['Ar condicionado', 'Sensor de estacionamento'...   
Aston Martin Virage    ['Travas elétricas', 'Controle de tração', 'Câ...   
Sandero                ['Piloto automático', 'Teto panorâmico', 'Câmb...   
XKR                    ['Controle de tração', 'Ar condicionado', 'Tra...   
X5                     ['Controle de tração', 'Piloto automático', 'T...   
Dodge Dakota           ['Controle de tração', 'Sensor de estacionamen...   
i30 CW                 ['Sensor de chuva', 'Vidros elétricos', 'Contr...   
Série 7 Sedã           ['Vidros elétricos', 'Travas elétricas', 'Roda...   
V60                    ['Rodas de liga', 'Sensor crepuscular', 'Ar co...   
Aston Martin DB5       ['Ar condicionado', '4

Vamos supor que queremos realizar uma query adicional, dessa vez verificando se esses veículos também são "Zero_km". Para isso, a partir do nosso dataset, passaremos entre parênteses dois pares de parênteses separados por um & (and, caractere conhecido como "e comercial").

In [122]:
dataset[(select) & (dataset.Zero_km==True)]


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Sorento              Motor Diesel  2019            NaN     True   
Cielo Hatch          Motor Diesel  2019            NaN     True   
Camry                Motor Diesel  2019            NaN     True   
Aston Martin Virage  Motor Diesel  2019            NaN     True   
Série 7 Sedã         Motor Diesel  2019            NaN     True   

                                                            Acessórios  \
Nome                                                                     
Sorento              ['Sensor de chuva', 'Câmera de estacionamento'...   
Cielo Hatch          ['Painel digital', 'Central multimídia', 'Câme...   
Camry                ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Aston Martin Virage  ['Travas elétricas', 'Controle de tração', 'Câ...   
Série 7 Sedã         ['Vidros elétricos', 'Travas elétricas', 'Roda...   

                         Valor  
Nome                            
Sorento               81399.35  
Cielo Hatch          145197.70  
Camry                138597.27  
Aston Martin Virage   97290.18  
Série 7 Sedã          67539.79

Existe outra forma de fazermos queries nos dataframes do Pandas, que é por meio da função query(). Com ela também podemos passar uma seleção, mas seguindo uma sintaxe um pouco diferente, na qual nossas comparações serão cercadas por aspas simples ou duplas. Note que se for necessário passar strings na seleção, elas deverão ser sinalizada pelo caractere que não foi utilizado na seleção superior - ou seja, se nossa seleção está cercada de aspas simples, a string deverá ser passada dentro de aspas duplas e vice-versa.

In [123]:
dataset.query('Motor=="Motor Diesel" and Zero_km == True')


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Sorento              Motor Diesel  2019            NaN     True   
Cielo Hatch          Motor Diesel  2019            NaN     True   
Camry                Motor Diesel  2019            NaN     True   
Aston Martin Virage  Motor Diesel  2019            NaN     True   
Série 7 Sedã         Motor Diesel  2019            NaN     True   

                                                            Acessórios  \
Nome                                                                     
Sorento              ['Sensor de chuva', 'Câmera de estacionamento'...   
Cielo Hatch          ['Painel digital', 'Central multimídia', 'Câme...   
Camry                ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Aston Martin Virage  ['Travas elétricas', 'Controle de tração', 'Câ...   
Série 7 Sedã         ['Vidros elétricos', 'Travas elétricas', 'Roda...   

                         Valor  
Nome                            
Sorento               81399.35  
Cielo Hatch          145197.70  
Camry                138597.27  
Aston Martin Virage   97290.18  
Série 7 Sedã          67539.79

Além do "and" (&), o operador "or" (|, "ou") também pode nos auxiliar nas nossas seleções.

In [124]:
dataset[(dataset.Motor == 'Motor Diesel') | (dataset.Zero_km == True)]


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Passat               Motor Diesel  1991         5712.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
A5                Motor 4.0 Turbo  2019            NaN     True   
J5                       Motor V6  2019            NaN     True   
A3                   Motor 1.0 8v  2019            NaN     True   
...                           ...   ...            ...      ...   
Benni Mini               Motor V8  2019            NaN     True   
Uno               Motor Diesel V6  2019            NaN     True   
Santa Fe            Motor 3.0 32v  2019            NaN     True   
XC60              Motor 4.0 Turbo  2019            NaN     True   
Aston Martin DB5     Motor Diesel  1996         7685.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
A5                ['Câmbio automático', 'Câmera de estacionament...   56445.20  
J5                ['Sensor crepuscular', 'Painel digital', 'Roda...   53183.38  
A3                ['4 X 4', 'Piloto automático', 'Central multim...   88552.39  
...                                                             ...        ...  
Benni Mini        ['Sensor crepuscular', 'Câmbio automático', 'C...  126247.84  
Uno               ['Central multimídia', 'Sensor crepuscular', '...  128852.21  
Santa Fe          ['Travas elétricas', 'Ar condicionado', '4 X 4...  129415.33  
XC60              ['Painel digital', 'Piloto automático', 'Centr...   77675.79  
Aston Martin DB5  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  

[82 rows x 6 columns]

In [125]:
# ou podemos fazer da seguinte forma:
dataset.query('Motor == "Motor Diesel" or Zero_km == True')


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Passat               Motor Diesel  1991         5712.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
A5                Motor 4.0 Turbo  2019            NaN     True   
J5                       Motor V6  2019            NaN     True   
A3                   Motor 1.0 8v  2019            NaN     True   
...                           ...   ...            ...      ...   
Benni Mini               Motor V8  2019            NaN     True   
Uno               Motor Diesel V6  2019            NaN     True   
Santa Fe            Motor 3.0 32v  2019            NaN     True   
XC60              Motor 4.0 Turbo  2019            NaN     True   
Aston Martin DB5     Motor Diesel  1996         7685.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
A5                ['Câmbio automático', 'Câmera de estacionament...   56445.20  
J5                ['Sensor crepuscular', 'Painel digital', 'Roda...   53183.38  
A3                ['4 X 4', 'Piloto automático', 'Central multim...   88552.39  
...                                                             ...        ...  
Benni Mini        ['Sensor crepuscular', 'Câmbio automático', 'C...  126247.84  
Uno               ['Central multimídia', 'Sensor crepuscular', '...  128852.21  
Santa Fe          ['Travas elétricas', 'Ar condicionado', '4 X 4...  129415.33  
XC60              ['Painel digital', 'Piloto automático', 'Centr...   77675.79  
Aston Martin DB5  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  

[82 rows x 6 columns]

EXERCÍCIOS  
  
Utilize o seguinte DataFrame para responder a questão:

In [127]:
import pandas as pd

dados = {
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8', 'Motor Diesel', 'Motor 1.6'],
    'Ano': [2019, 2003, 1991, 2019, 1990],
    'Quilometragem': [0.0, 5712.0, 37123.0, 0.0, 120000.0],
    'Zero_km': [True, False, False, True, False],
    'Valor': [88000.0, 106000.0, 72000.0, 89000.0, 32000.0]
}

datasetexercicio = pd.DataFrame(dados, index = ['Jetta', 'Passat', 'Crossfox', 'DS5', 'Fusca'])


In [131]:
datasetexercicio.query('Motor =="Motor Diesel" or Zero_km == True')


Motor   Ano  Quilometragem  Zero_km     Valor
Jetta   Motor 4.0 Turbo  2019            0.0     True   88000.0
Passat     Motor Diesel  2003         5712.0    False  106000.0
DS5        Motor Diesel  2019            0.0     True   89000.0

In [132]:
# Também poderia fazer da seguinte forma:
datasetexercicio.query('Motor == "Motor Diesel" | Zero_km == True' )


Motor   Ano  Quilometragem  Zero_km     Valor
Jetta   Motor 4.0 Turbo  2019            0.0     True   88000.0
Passat     Motor Diesel  2003         5712.0    False  106000.0
DS5        Motor Diesel  2019            0.0     True   89000.0

5.4 Iterando com DataFrames

In [133]:
dataset.head()


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

In [134]:
# não é a forma correta!
for item in dataset:
    print(item)


Motor
Ano
Quilometragem
Zero_km
Acessórios
Valor


A forma abaixo para iterar dentro de um DataFrame serve para bases pequenas... não é ideal para bases grandes de dados

In [139]:
# Forma correta - Desempacotar as tuplas 
for index, row in dataset.iterrows(): # itera pelas linhas do índice
    if(2019 - row['Ano'] != 0): # se a linha do ano for diferente de 0
        dataset.loc[index, 'km_media'] = row['Quilometragem'] / (2019 - row['Ano']) # localize o indice e inclua uma coluna com o nome de 'km_media' e faça o cálculo desta linha que é a quilometragem desta linha, dividido pelo ano atual (2019) - o Ano dessa linha
    else: dataset.loc[index,'km_media']=0 # se não atribua o valor = 0 / para a nova coluna 'km_media'.
dataset


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor     km_media  
Nome                                           
Jetta Variant           88078.64  2775.625000  
Passat                 106161.94   204.000000  
Crossfox                72832.16  1280.103448  
DS5                    124549.07     0.000000  
Aston Martin DB4        92612.10  1981.307692  
...                          ...          ...  
Phantom 2013            51759.58  5501.000000  
Cadillac Ciel concept   51667.06  1070.750000  
Classe GLK              68934.03  3096.294118  
Aston Martin DB5       122110.90   334.130435  
Macan                   90381.47  1858.814815  

[258 rows x 7 columns]

5.5 TRATAMENTOS DE DADOS

In [140]:
dataset.head()


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios  \
Nome                                                                  
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox          ['Piloto automático', 'Controle de estabilidad...   
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   

                      Valor     km_media  
Nome                                      
Jetta Variant      88078.64  2775.625000  
Passat            106161.94   204.000000  
Crossfox           72832.16  1280.103448  
DS5               124549.07     0.000000  
Aston Martin DB4   92612.10  1981.307692

In [142]:
# Coletando informações do DataSet
dataset.info()


<class 'pandas.core.frame.DataFrame'>
Index: 258 entries, Jetta Variant to Macan
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Motor          258 non-null    object 
 1   Ano            258 non-null    int64  
 2   Quilometragem  197 non-null    float64
 3   Zero_km        258 non-null    bool   
 4   Acessórios     258 non-null    object 
 5   Valor          258 non-null    float64
 6   km_media       258 non-null    float64
dtypes: bool(1), float64(3), int64(1), object(2)
memory usage: 22.5+ KB


In [143]:
# O Pandas nos oferece a possibilidade de selecionar informações nulas ou não: a função isna(). Ela traz um retorno boleano
dataset.Quilometragem.isna()


Nome
Jetta Variant            False
Passat                   False
Crossfox                 False
DS5                       True
Aston Martin DB4         False
                         ...  
Phantom 2013             False
Cadillac Ciel concept    False
Classe GLK               False
Aston Martin DB5         False
Macan                    False
Name: Quilometragem, Length: 258, dtype: bool

In [144]:
dataset[dataset.Quilometragem.isna()]

# O resultado nos traz a nítida percepção de que os carros que possuem NaN, são na verdade Zero_km = 0


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                  
DS5                   Motor 2.4 Turbo  2019            NaN     True   
A5                    Motor 4.0 Turbo  2019            NaN     True   
J5                           Motor V6  2019            NaN     True   
A3                       Motor 1.0 8v  2019            NaN     True   
Série 1 M                    Motor V8  2019            NaN     True   
...                               ...   ...            ...      ...   
Lamborghini Reventón  Motor 4.0 Turbo  2019            NaN     True   
Benni Mini                   Motor V8  2019            NaN     True   
Uno                   Motor Diesel V6  2019            NaN     True   
Santa Fe                Motor 3.0 32v  2019            NaN     True   
XC60                  Motor 4.0 Turbo  2019            NaN     True   

                                                             Acessórios  \
Nome                                                                      
DS5                   ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
A5                    ['Câmbio automático', 'Câmera de estacionament...   
J5                    ['Sensor crepuscular', 'Painel digital', 'Roda...   
A3                    ['4 X 4', 'Piloto automático', 'Central multim...   
Série 1 M             ['Controle de estabilidade', 'Central multimíd...   
...                                                                 ...   
Lamborghini Reventón  ['Controle de tração', 'Ar condicionado', 'Cen...   
Benni Mini            ['Sensor crepuscular', 'Câmbio automático', 'C...   
Uno                   ['Central multimídia', 'Sensor crepuscular', '...   
Santa Fe              ['Travas elétricas', 'Ar condicionado', '4 X 4...   
XC60                  ['Painel digital', 'Piloto automático', 'Centr...   

                          Valor  km_media  
Nome                                       
DS5                   124549.07       0.0  
A5                     56445.20       0.0  
J5                     53183.38       0.0  
A3                     88552.39       0.0  
Série 1 M              94564.40       0.0  
...                         ...       ...  
Lamborghini Reventón   67664.86       0.0  
Benni Mini            126247.84       0.0  
Uno                   128852.21       0.0  
Santa Fe              129415.33       0.0  
XC60                   77675.79       0.0  

[61 rows x 7 columns]

In [145]:
# Substituindo o NaN por 0 /  fillna preenche -- Atenção que esta função só faz a alteração na visualização dos dados e não no DataSet
dataset.fillna(0)


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            0.0     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor     km_media  
Nome                                           
Jetta Variant           88078.64  2775.625000  
Passat                 106161.94   204.000000  
Crossfox                72832.16  1280.103448  
DS5                    124549.07     0.000000  
Aston Martin DB4        92612.10  1981.307692  
...                          ...          ...  
Phantom 2013            51759.58  5501.000000  
Cadillac Ciel concept   51667.06  1070.750000  
Classe GLK              68934.03  3096.294118  
Aston Martin DB5       122110.90   334.130435  
Macan                   90381.47  1858.814815  

[258 rows x 7 columns]

In [146]:
# Para que a alteração seja efetivada no DataSe, é necessário acrescentar mais um parâmentro = inplace = True
dataset.fillna(0, inplace = True)


In [147]:
dataset


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            0.0     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor     km_media  
Nome                                           
Jetta Variant           88078.64  2775.625000  
Passat                 106161.94   204.000000  
Crossfox                72832.16  1280.103448  
DS5                    124549.07     0.000000  
Aston Martin DB4        92612.10  1981.307692  
...                          ...          ...  
Phantom 2013            51759.58  5501.000000  
Cadillac Ciel concept   51667.06  1070.750000  
Classe GLK              68934.03  3096.294118  
Aston Martin DB5       122110.90   334.130435  
Macan                   90381.47  1858.814815  

[258 rows x 7 columns]

In [148]:
# Quero verificar se agora, todo veículo zero está tratado, ou seja, com 0.0 ao invés de NaN
dataset.query("Zero_km == True")


Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                  
DS5                   Motor 2.4 Turbo  2019            0.0     True   
A5                    Motor 4.0 Turbo  2019            0.0     True   
J5                           Motor V6  2019            0.0     True   
A3                       Motor 1.0 8v  2019            0.0     True   
Série 1 M                    Motor V8  2019            0.0     True   
...                               ...   ...            ...      ...   
Lamborghini Reventón  Motor 4.0 Turbo  2019            0.0     True   
Benni Mini                   Motor V8  2019            0.0     True   
Uno                   Motor Diesel V6  2019            0.0     True   
Santa Fe                Motor 3.0 32v  2019            0.0     True   
XC60                  Motor 4.0 Turbo  2019            0.0     True   

                                                             Acessórios  \
Nome                                                                      
DS5                   ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
A5                    ['Câmbio automático', 'Câmera de estacionament...   
J5                    ['Sensor crepuscular', 'Painel digital', 'Roda...   
A3                    ['4 X 4', 'Piloto automático', 'Central multim...   
Série 1 M             ['Controle de estabilidade', 'Central multimíd...   
...                                                                 ...   
Lamborghini Reventón  ['Controle de tração', 'Ar condicionado', 'Cen...   
Benni Mini            ['Sensor crepuscular', 'Câmbio automático', 'C...   
Uno                   ['Central multimídia', 'Sensor crepuscular', '...   
Santa Fe              ['Travas elétricas', 'Ar condicionado', '4 X 4...   
XC60                  ['Painel digital', 'Piloto automático', 'Centr...   

                          Valor  km_media  
Nome                                       
DS5                   124549.07       0.0  
A5                     56445.20       0.0  
J5                     53183.38       0.0  
A3                     88552.39       0.0  
Série 1 M              94564.40       0.0  
...                         ...       ...  
Lamborghini Reventón   67664.86       0.0  
Benni Mini            126247.84       0.0  
Uno                   128852.21       0.0  
Santa Fe              129415.33       0.0  
XC60                   77675.79       0.0  

[61 rows x 7 columns]

In [149]:
# No decorrer da evolução dos estudos, tivemos o data Set Modificado, para trazer ele originalmente, temos que importá-lo novamente.
dataset = pd.read_csv(r"C:\Users\fabia\OneDrive\FIAP_INICIO\WarnUp_materiais\Python_Data_Science\Pandas\data\db.csv", sep = ';')


In [150]:
dataset


Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
3                      DS5        Motor 2.4 Turbo  2019            NaN   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
3       True  ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[258 rows x 7 columns]

In [151]:
# Também posso eliminar os casos NaN caso seja necessário com o 'drona'
dataset.dropna(subset=['Quilometragem'], inplace=True)


In [152]:
dataset


Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
5            Palio Weekend          Motor 1.8 16v  2012        10728.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
5      False  ['Sensor de estacionamento', 'Teto panorâmico'...   97497.73  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[197 rows x 7 columns]

Feito isso, nosso conjunto passará a ter somente 197 registros, diferente dos 258 anteriores. Aprendemos então duas abordagens diferentes para lidarmos com dados problemáticos nos nossos conjuntos: preencher os registros com alguma informação que faça mais sentido ou seja útil, ou simplesmente eliminá-los.

Nesta aula, aprendemos:

* Series e DataFrame, as estruturas de dados básicas da biblioteca pandas
* O método de criação de Series e DataFrames
* A criar DataFrames a partir de dados externos
* Como fazer seleções e fatiamentos em um DataFrame
* Os métodos de seleção .loc e .iloc
* A fazer consultas em um DataFrame com o método .query() e com matrizes booleanas
* A forma básica de iteração por um DataFrame
* O básico sobre tratamento de dados faltantes em um DataFrame (isna(), fillna() e dropna())